### PR TITLE
[platform_tests/counterpoll]: Support multi-asic for counterpoll watermark tests

### LINKED ISSUE #14753: [Bug]: counterpoll watermark test does not support multi-asic
### Issue Description

counterpoll watermark tests fail on multi-asic linecards as they do not look at the namespaced database containers. Refactor the tests to support multi-asic

### Results you see

COUNTER_DB is not detected as populated in multi asic devices because the test does not look at the database container for each asic

### Results you expected to see

COUNTER_DB is populated for each individual linecard, test should look at the proper database container

### Is it platform specific

generic

### Relevant log output

_No response_

### Output of `show version`

_No response_

### Attach files (if any)

_No response_

### PR DESCRIPTION
**Pull request by liamkearney-msft**
* Update counterpoll watermark platform test to support multi-asic DUTs

BugLink: https://github.com/sonic-net/sonic-mgmt/issues/14753

### Description of PR

Refactor test_counterpoll_watermark to properly handle multi-asic DUTs. The previous version of the test would fail on multi-asic as it would not look at the namespaced database containers.

Summary:
Fixes #14753 

### Type of change

- [x ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ x] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405

#### What is the motivation for this PR?
Fix test case to properly run on multi-asic
#### How did you verify/test it?
Ran the test on a testbed with multi-asic and single asic devices. Test runs without errors in testplan logic

### CODE REVIEW

**Review by auspham — APPROVED** [review]
LGTM

File: tests/platform_tests/counterpoll/counterpoll_helper.py [hunk 1]
```diff
@@ -1,9 +1,12 @@
 from tests.platform_tests.counterpoll.counterpoll_constants import CounterpollConstants
+from tests.common.devices.sonic_asic import SonicAsic
 
 
 class ConterpollHelper:
     @staticmethod
     def get_counterpoll_show_output(duthost):
+        if isinstance(duthost, SonicAsic):
```
Comment (arlakshm) [inline tests/platform_tests/counterpoll/counterpoll_helper.py:8]: can we implement `show_and_parse` in `SonicAsic` we can avoid checking duthost is instance of `SonicAsic` or not

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py [hunk 2]
```diff
@@ -237,14 +258,15 @@ def verify_counterpoll_status(duthost, counterpoll_list, expected):
...
-        pytest_assert(v > 0, "watermark_stats {} in COUNTERS_DB: {}, expected > 0".format(k, v))
+    watermark_stats_output = get_keys_on_asics(duthost, 'COUNTERS_DB', '*{}*'
+                                               .format(CounterpollConstants.WATERMARK.upper()))
+    for asic_idx, output in watermark_stats_output.items():
+        watermark_stats = {}
+        for watermark_type in WATERMARK_COUNTERS_DB_STATS_TYPE:
+            watermark_stats[watermark_type] = 0
+            for line in output:
+                if watermark_type in line:
+                    watermark_stats[watermark_type] += 1
```
Comment (arlakshm) [inline tests/platform_tests/counterpoll/test_counterpoll_watermark.py:269]: The test case will not detect if one of the asic does not have any watermark stat?

**Review by arlakshm — COMMENTED** [review]
can you please test this change on T0/T1 testbed to make sure these changes do not break single ASIC platforms

File: tests/platform_tests/counterpoll/test_counterpoll_watermark.py (hunk 2 above)
Comment (liamkearney-msft) [reply tests/platform_tests/counterpoll/test_counterpoll_watermark.py:269]: it iterates through each asic one at a time - if it finds one where there is missing watermarks itll assert and stop.

File: tests/platform_tests/counterpoll/counterpoll_helper.py (hunk 1 above)
Comment (liamkearney-msft) [reply tests/platform_tests/counterpoll/counterpoll_helper.py]: done

**Comment by liamkearney-msft** [comment]
@arlakshm tested on t0 TB, and t2 TB with a multi-asic and single-asic linecard - tests pass

**Comment by auspham** [comment]
Hi @yejianquan  could you help reviewing this PR and merge for next nightly run. Thank you

**Review by arlakshm — APPROVED** [review]

**Pull request merged**